tracer/test/Datadog.Trace.Tests/Sampling/SpanSamplerTests.cs

```diff
@@ -0,0 +1,210 @@
+            Assert.Null(span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism));
+        }
+
+        [Fact]
+        public void MaxPerSecond_ShouldNotBeTagged_WhenNotExists()
+        {
+            var expectedRuleRate = 0.99f;
+            var expectedSamplingMechanism = 8; // SamplingMechanism.SpanSamplingRule
+            var rule1 = new SpanSamplingRule("service-name", "operation-name", 0.99f);
+            var rules = new List<SpanSamplingRule>() { rule1 };
+            var sampler = new SpanSampler(rules);
+            var span = new Span(new SpanContext(5, 6, null, serviceName: "service-name"), DateTimeOffset.Now) { OperationName = "operation-name" };
+
+            sampler.MakeSamplingDecision(span);
+
+            Assert.Equal(expectedRuleRate.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.RuleRate));
+            Assert.Null(span.Tags.GetTag(Tags.SingleSpanSampling.MaxPerSecond));
+            Assert.Equal(expectedSamplingMechanism.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism));
+        }
+
+        [Fact]
+        public void Tags_ShouldBe_Correct()
+        {
+            var expectedRuleRate = 0.99f;
+            var expectedMaxPerSecond = 500.0f;
+            var expectedSamplingMechanism = 8; // SamplingMechanism.SpanSamplingRule
+            var rule1 = new SpanSamplingRule("service-name", "operation-name", 0.99f, 500.0f);
+            var rules = new List<SpanSamplingRule>() { rule1 };
+            var sampler = new SpanSampler(rules);
+            var span = new Span(new SpanContext(5, 6, null, serviceName: "service-name"), DateTimeOffset.Now) { OperationName = "operation-name" };
+
+            sampler.MakeSamplingDecision(span);
+
+            Assert.Equal(expectedRuleRate.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.RuleRate));
+            Assert.Equal(expectedMaxPerSecond.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.MaxPerSecond));
+            Assert.Equal(expectedSamplingMechanism.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism));
+        }
+
+        [Fact]
+        public void Allow_All()
+        {
+            var allowAllRule = new SpanSamplingRule("*", "*");
+            var rules = new List<SpanSamplingRule>() { allowAllRule };
+            var sampler = new SpanSampler(rules);
+
+            RunSamplerTest(sampler, 500, expectedAutoKeepRate: 0, expectedUserKeepRate: 1, acceptableVariancePercent: 0);
+        }
+
+        [Fact]
+        public void Allow_None_SamplingRate()
+        {
+            var allowNoneRule = new SpanSamplingRule("*", "*", 0.0f);
+            var rules = new List<SpanSamplingRule>() { allowNoneRule };
+            var sampler = new SpanSampler(rules);
+
+            RunSamplerTest(sampler, 500, expectedAutoKeepRate: 1, expectedUserKeepRate: 0, acceptableVariancePercent: 0);
+        }
+
+        [Fact]
+        public void Allow_Half_SamplingRate()
```

Comment:
You should maybe add a test with a `?` in the service name or operation name pattern.

Reply:
Added some more to this in f7844237db564d604c188c106c384ed644307fe6